tools/ci-watch/main.go

```diff
@@ -0,0 +1,343 @@
+// Copyright 2025 Edgeless Systems GmbH
+// SPDX-License-Identifier: BUSL-1.1
+
+package main
+
+import (
+	"archive/zip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-github/v72/github"
+)
+
+func main() {
+	if len(os.Args) < 2 {
+		fmt.Println("Usage: ci-watch <run-url>")
+		os.Exit(1)
+	}
+
+	runURL := os.Args[1]
+	u, err := url.Parse(runURL)
+	if err != nil {
+		fmt.Printf("Invalid URL: %v\n", err)
+		os.Exit(1)
+	}
+
+	// Expected format: /owner/repo/actions/runs/runID
+	parts := strings.Split(u.Path, "/")
+	if len(parts) < 6 {
+		fmt.Println("Invalid URL format. Expected: https://github.com/owner/repo/actions/runs/runID")
+		os.Exit(1)
+	}
+
+	var owner, repo, runID string
+	// Find "actions" and "runs" to locate IDs
+	for i, part := range parts {
+		if part == "actions" && i+2 < len(parts) && parts[i+1] == "runs" {
+			runID = parts[i+2]
+			if i >= 2 {
+				repo = parts[i-1]
+				owner = parts[i-2]
+			}
+			break
+		}
+	}
+
+	if owner == "" || repo == "" || runID == "" {
+		fmt.Println("Could not parse owner, repo, or runID from URL")
+		os.Exit(1)
+	}
+
+	targetDir := "logs"
+	fmt.Printf("Fetching logs for %s/%s run %s to %s\n", owner, repo, runID, targetDir)
+
+	if err := fetchGithubWorkflowRun(context.Background(), owner, repo, runID, targetDir); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+	fmt.Println("Done")
+}
+
+// fetchGithubWorkflowRun fetches the logs and artifacts of a GitHub Actions workflow run.
+func fetchGithubWorkflowRun(ctx context.Context, owner, repo, runID, targetDir string) error {
+	token := os.Getenv("GITHUB_TOKEN")
+	if token == "" {
+		return fmt.Errorf("GITHUB_TOKEN environment variable is not set")
+	}
+
+	runIDInt, err := strconv.ParseInt(runID, 10, 64)
+	if err != nil {
+		return fmt.Errorf("invalid run ID %q: %w", runID, err)
+	}
+
+	client := github.NewClient(nil).WithAuthToken(token)
+
+	// Create run directory
+	runDir := filepath.Join(targetDir, runID)
+	if err := os.MkdirAll(runDir, 0o755); err != nil {
+		return fmt.Errorf("failed to create run directory: %w", err)
+	}
+
+	var errs []error
+
+	fmt.Println("Fetching job console logs...")
+	if err := fetchJobConsoleLogs(ctx, client, owner, repo, runIDInt, runDir); err != nil {
+		errs = append(errs, fmt.Errorf("failed to fetch job logs: %w", err))
+		fmt.Fprintf(os.Stderr, "Error fetching job logs: %v\n", err)
+	}
+
+	fmt.Println("Fetching artifacts...")
+	if err := fetchRunArtifacts(ctx, client, owner, repo, runIDInt, runDir); err != nil {
+		errs = append(errs, fmt.Errorf("failed to fetch artifacts: %w", err))
+		fmt.Fprintf(os.Stderr, "Error fetching artifacts: %v\n", err)
+	}
+
+	return errors.Join(errs...)
```

Comment:
For general failures this will still iterate over everything and return the same error over and over

Reply:
sounds okay for now, this is just dev tooling.